pkg/upgrade/control_plane.go

```diff
@@ -343,144 +395,245 @@ func (u *ControlPlaneUpgrader) updateMachine(name string, machine clusterv1.Mach
 	if err != nil {
 		return err
 	}
 
-	mo := MachineOptions{
-		ImageID:        u.imageID,
-		ImageField:     u.imageField,
-		DesiredVersion: u.desiredVersion,
-	}
+	for _, machine := range machines {
+		log := u.log.WithValues(
+			"machine", fmt.Sprintf("%s/%s", machine.Namespace, machine.Name),
+			"upgrade-id", u.upgradeID,
+		)
 
-	machineCreator := NewMachineCreator(
-		WithManagementClient(u.managementClusterClient),
-		WithNodeLister(u.targetKubernetesClient.CoreV1().Nodes()),
-		WithPodGetter(u.targetKubernetesClient.CoreV1().Pods("kube-system")),
-		WithMachineOptions(mo),
-		WithLogger(u.log.WithName("machine-creator")),
-	)
+		if machine.Spec.ProviderID == nil {
+			log.Info("unable to upgrade machine as it has no spec.providerID")
+			// TODO record event/annotation?
+			continue
+		}
 
-	// TODO add more error logs on failure conditions
-	for _, machine := range machines.Items {
 		annotations := machine.GetAnnotations()
-		// Skip any machine that already has the annotation we're looking for
-		if val, ok := annotations[UpgradeIDAnnotationKey]; ok && val == u.upgradeID {
+		if annotations == nil {
+			annotations = make(map[string]string)
+			machine.SetAnnotations(annotations)
+		}
+
+		// Add upgrade ID if it isn't there
+		if annotations[AnnotationUpgradeID] == "" {
+			helper, err := patch.NewHelper(machine.DeepCopy(), u.managementClusterClient)
+			if err != nil {
+				// TODO should we do anything else?
+				log.Error(err, "error creating patch helper for machine (add upgrade id)")
+				continue
+			}
+
+			machine.Annotations[AnnotationUpgradeID] = u.upgradeID
+
+			log.Info("Storing upgrade ID on machine")
+
+			if err := helper.Patch(context.TODO(), machine); err != nil {
+				// TODO should we do anything else?
+				log.Error(err, "error patching machine (add upgrade id)")
+				continue
+			}
+		}
+
+		// Don't process a mismatching upgrade ID
+		if annotations[AnnotationUpgradeID] != u.upgradeID {
+			// TODO record that we're unable to upgrade because the ID is a mismatch (annotation? event?)
+			log.Info("Unable to upgrade machine - mismatching upgrade id", "machine-upgrade-id", annotations[AnnotationUpgradeID])
 			continue
 		}
 
-		if machine.Spec.ProviderID == nil {
-			u.log.Info("unable to upgrade machine as it has no spec.providerID", "name", machine.Name)
+		// Skip if this is a replacement machine for the current upgrade
+		if strings.HasSuffix(machine.Name, upgradeSuffix(u.upgradeID)) {
+			log.Info("Skipping machine as it is a replacement machine for the in-process upgrade")
 			continue
 		}
 
-		// assume the original name is controlplane-<index> or controlplane-<index>-<timestamp>
-		// let's set the new name to controlplane-<index>-<timestamp>
-		nameParts := strings.Split(machine.Name, "-")
-		if len(nameParts) < 2 {
-			return errors.Errorf("machine name %q does not match expected format <name>-<index> or <name>-<index>-<timestamp>", machine.Name)
+		// Skip if we've already completed this machine
```

Comment:
I'm not sure if this is relevant. If we're upgrading an original machine _O_, it will have the upgrade-id annotation, but we only ever apply the completed annotation to the replacement machine _R_. When we complete the upgrade of _O_, we delete it. When we complete the upgrade of the entire control plane, we remove all upgrade-tool annotations from all the replacement machines. WDYT @chuckha?

Reply:
hmm, I was thinking we'd apply the completed annotations to the original machine so we could skip it at this point since we already filter out replacement machines from "list of machines that need upgrades"

Reply:
Given that everything is idempotent now, we'd be inserting a step to annotate the machine as completed in between deleting it from etcd, and deleting the machine itself. I don't think it's worth it? If we don't manage to delete a machine but everything else before that succeeds, the next time we run the tool, it'd be a no-op until we get to the machine deletion part. Right?

Reply:
ugh, my fault. This stuff is hard to rebuild context around.

Yeah, you're right, I agree this section isn't relevant then.